src/test/java/org/springframework/data/mapping/PreferredConstructorDiscovererUnitTests.java

```diff
@@ -149,6 +150,34 @@ void detectsMetaAnnotatedValueAnnotation() {
 		});
 	}
 
+	@Test // GH-2332
+	void detectsCanonicalRecordConstructor() {
+
+		assertThat(PreferredConstructorDiscoverer.discover(RecordWithSingleArgConstructor.class)).satisfies(ctor -> {
+
+			assertThat(ctor.getParameters()).hasSize(2);
+			assertThat(ctor.getParameters().get(0).getRawType()).isEqualTo(Long.class);
+			assertThat(ctor.getParameters().get(1).getRawType()).isEqualTo(String.class);
+		});
+
+		assertThat(PreferredConstructorDiscoverer.discover(RecordWithNoArgConstructor.class)).satisfies(ctor -> {
+
+			assertThat(ctor.getParameters()).hasSize(2);
+			assertThat(ctor.getParameters().get(0).getRawType()).isEqualTo(Long.class);
+			assertThat(ctor.getParameters().get(1).getRawType()).isEqualTo(String.class);
+		});
+	}
+
+	@Test // GH-2332
+	void detectsAnnotatedRecordConstructor() {
+
+		assertThat(PreferredConstructorDiscoverer.discover(RecordWithPersistenceCreator.class)).satisfies(ctor -> {
```

Comment:
Can we make these either soft asserts, or separate tests?

Reply:
For whatever reason, the entire test class is written in that way. While I find that style strange (likely that comes from our `Optional` experiments), we could apply some polishing post-RC1.